File: 30-Day LeetCoding Challenge/Python/Day 3 - Maximum Subarray.py
import sys
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        global_max = - sys.maxsize - 1
        local_max = 0
        for num in nums:
            local_max = max(num, local_max + num)
            global_max = max(local_max, global_max)
        return global_max

nums = [-2,1,-3,4,-1,2,1,-5,4]
solution = Solution()
print(solution.maxSubArray(nums))

        

File: 30-Day LeetCoding Challenge/C++/Day 3 - Maximum Subarray.cpp
#include <iostream>
#include <vector>
#include "limits.h"
using namespace std;

class Solution {
    public:
        int maxSubArray(vector<int> &nums) {
            int global_max = INT_MIN;
            int local_max = 0;
            for(int num : nums) {
                local_max = max(num, local_max + num);
                global_max = max(local_max, global_max);
            }
            return global_max;
        }
};
int main() {
    Solution s;
    vector<int> nums{-2,1,-3,4,-1,2,1,-5,4};
    cout << s.maxSubArray(nums);
}